15 спринт

David-Arcanone opened the pull request.

David-Arcanone committed:
задание спринта 15
разделил на data, domain, presentation
функцияю dpToPx отдельно вынес в Utilities, тк он может быть полезен в любом UI слое гдя я буду использовать постеры.

David-Arcanone committed:
задание спринта 15
разделил на data, domain, presentation
функцияю dpToPx отдельно вынес в Utilities, тк он может быть полезен в любом UI слое гдя я буду использовать постеры.

David-Arcanone committed:
задание спринта 15
разделил на data, domain, presentation
функцияю dpToPx отдельно вынес в Utilities, тк он может быть полезен в любом UI слое гдя я буду использовать постеры.

kekulta commented on app/build.gradle.kts:
```diff
@@ -26,6 +26,9 @@ android {
             )
         }
     }
+    buildFeatures {
+        viewBinding = true
+    }
```
Молодец, что перешёл на viewBinding!

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/settings/SettingsActivity.kt:
```diff
@@ -0,0 +1,41 @@
+
+class SettingsActivity : AppCompatActivity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        val mySettingsInteractorImpl = Creator.provideGetSettingsInteractor(this)
```
Нельзя провайдить куда-то контекст из Активити, есть риск того, что он может утечь.
Правильным решением было бы сделать метод вроде `setContext(Context)` у Creator и инициализировать его с помощью ApplicationContext в onCreate класса app.

David-Arcanone replied:
сделал через App. Скорее всего моя доп проверка внутри Creator лишняя. 
Если такая реализация устраивает, инициализирую SharedPreference там же в Creator и раздам моим Репозиториям.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/settings/SettingsActivity.kt:
```diff
@@ -0,0 +1,41 @@
+        setContentView(R.layout.activity_settings)
+        //кнопка назад
+        val btBack = findViewById<Button>(R.id.bt_back)
+        btBack.setOnClickListener {
+            val backIntent = Intent(this, MainActivity::class.java)
+            startActivity(backIntent)
+        }
```
Так делать неправильно. Когда ты создаёшь новую активити, то она добавляется в стек поверх старой.
Если ты будешь открывать и закрывать этот экран, то ты будешь поочерёдно класть в стек главный и этот экраны.
Стек будет расти и в какой-то момент это может привести к проблемам с памятью.
Правильный подход закрывать активити, когда уходишь с экран, например, с помощью метод `finish()`.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/settings/SettingsActivity.kt:
```diff
@@ -0,0 +1,41 @@
+            startActivity(backIntent)
+        }
+        //кнопка поделиться
+        val btShare = findViewById<Button>(R.id.bt_share)
+        btShare.setOnClickListener { mySettingsInteractorImpl.openShare() }
+        //кнопка поддержка
+        val btSupport = findViewById<Button>(R.id.bt_support)
+        btSupport.setOnClickListener { mySettingsInteractorImpl.openSupport() }
+        //кнопка соглашения об использовании
+        val btUserAgreement = findViewById<Button>(R.id.bt_user_agreement)
+        btUserAgreement.setOnClickListener { mySettingsInteractorImpl.openUserAgreement() }
+        //поменять тему
+        val themeSwitcher = findViewById<SwitchMaterial>(R.id.themeSwitcher)
+        val savedNightTheme = mySettingsInteractorImpl.getSavedNightTheme()
```
Вижу, что выше включил viewBinding, но тут им не пользуешься.

David-Arcanone replied:
я собирался, но в след теме как раз ViewBiding, решил сперва с этим разобраться.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/search/SearchActivity.kt:
```diff
@@ -19,16 +19,28 @@ import android.widget.TextView
 
 class SearchActivity : AppCompatActivity() {
+    //инициализация необходимых переменных
+    private lateinit var myITunesInteractorImpl: SearchInteractor
```
impl не очень правильное название переменной, мы работаем только с интерфейсами.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/utils/Utilities.kt:
```diff
@@ -1,10 +1,11 @@
 
 object Utilities {
-    val iTunesBaseUrl = "https://itunes.apple.com"
+    const val PLAYLIST_SAVED_PREFERENCES = "playlist_saved_preferences"
+    const val NIGHT_THEME_KEY = "key_for_night_theme"
     fun dpToPx(dp: Float, context: Context): Int {
         return TypedValue.applyDimension(
             TypedValue.COMPLEX_UNIT_DIP,
```
Я бы разделил Utils на два.
Есть AndroidUtils, которые используют контекст и внешние зависимости и ими можно пользоваться в слоях presentation и data. 
Есть просто Utils, которые не используют внешних зависимостей и ими можно пользоваться так же и в домейне.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/domain/api/SearchInteractor.kt:
```diff
@@ -0,0 +1,17 @@
+interface SearchInteractor {
+    fun searchTracks(expression: String, consumer: TracksConsumer<SearchedTracks>)
+    fun getSavedTrackHistory(): ArrayList<Track>
+    fun saveTrackHistory(newValue: ArrayList<Track>)
```
Вместо имплементации ArrayList лучше использовать интерфейс List.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/domain/api/SearchInteractor.kt:
```diff
@@ -0,0 +1,17 @@
+    fun getSavedTrack(): Track?
+    fun saveCurrentTrack(newValue: Track)
+    fun getSavedInstanceEditTextValue(savedInstanceState: Bundle?): String
+    fun saveInstanceEditTextValue(outState: Bundle, myText: String)
```
Bundle это класс из Android SDK, его не может быть в домейне.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/data/repository/PlayerRepositoryImpl.kt:
```diff
@@ -0,0 +1,50 @@
+    private val sharedPref: SharedPreferences,
+    private val mediaPlayer: MediaPlayer
+) : PlayerRepository {
+    companion object {
+        private const val PLAYLIST_CURRENT_TRACK = "playlist_current_track"
+    }
+
+    override fun getSavedTrack(): Track? {
```
Согласно [код-стайлу](https://kotlinlang.org/docs/coding-conventions.html#class-layout) Котлина Companion Object должен находится в конце класса.
Хотя я лично знаю команды, где он пишется в начале класса.
В целом, код-стайл это не про жёсткие правила, а про договорённости внутри
команды. Но я бы рекомендовал использовать стандартные практики в новых
проектах, где ещё не устоялись нестадартные. Это будет очень полезно при
онбординге новых разработчиков.

В целом рекомендую внимательно ознакомиться с Coding Conventions Котлина и по
возможности им следовать.

David-Arcanone committed:
задание спринта 15
согласно совету сделал функцию для setContext в onCreate App. Если реализация устраивает ревьювера уберу запросы SharedPreference со всех Activity (не сделал сейчас на случай если придется переделывать)

David-Arcanone committed:
задание спринта 15
больше не надо брать sharedPreferences в каждом активити, беру его только 1 раз
